<html>
    <head>
        <meta charset="utf-8">
        
            <script>function neighbourhoodHighlight(params) {
  // console.log("in nieghbourhoodhighlight");
  allNodes = nodes.get({ returnType: "Object" });
  // originalNodes = JSON.parse(JSON.stringify(allNodes));
  // if something is selected:
  if (params.nodes.length > 0) {
    highlightActive = true;
    var i, j;
    var selectedNode = params.nodes[0];
    var degrees = 2;

    // mark all nodes as hard to read.
    for (let nodeId in allNodes) {
      // nodeColors[nodeId] = allNodes[nodeId].color;
      allNodes[nodeId].color = "rgba(200,200,200,0.5)";
      if (allNodes[nodeId].hiddenLabel === undefined) {
        allNodes[nodeId].hiddenLabel = allNodes[nodeId].label;
        allNodes[nodeId].label = undefined;
      }
    }
    var connectedNodes = network.getConnectedNodes(selectedNode);
    var allConnectedNodes = [];

    // get the second degree nodes
    for (i = 1; i < degrees; i++) {
      for (j = 0; j < connectedNodes.length; j++) {
        allConnectedNodes = allConnectedNodes.concat(
          network.getConnectedNodes(connectedNodes[j])
        );
      }
    }

    // all second degree nodes get a different color and their label back
    for (i = 0; i < allConnectedNodes.length; i++) {
      // allNodes[allConnectedNodes[i]].color = "pink";
      allNodes[allConnectedNodes[i]].color = "rgba(150,150,150,0.75)";
      if (allNodes[allConnectedNodes[i]].hiddenLabel !== undefined) {
        allNodes[allConnectedNodes[i]].label =
          allNodes[allConnectedNodes[i]].hiddenLabel;
        allNodes[allConnectedNodes[i]].hiddenLabel = undefined;
      }
    }

    // all first degree nodes get their own color and their label back
    for (i = 0; i < connectedNodes.length; i++) {
      // allNodes[connectedNodes[i]].color = undefined;
      allNodes[connectedNodes[i]].color = nodeColors[connectedNodes[i]];
      if (allNodes[connectedNodes[i]].hiddenLabel !== undefined) {
        allNodes[connectedNodes[i]].label =
          allNodes[connectedNodes[i]].hiddenLabel;
        allNodes[connectedNodes[i]].hiddenLabel = undefined;
      }
    }

    // the main node gets its own color and its label back.
    // allNodes[selectedNode].color = undefined;
    allNodes[selectedNode].color = nodeColors[selectedNode];
    if (allNodes[selectedNode].hiddenLabel !== undefined) {
      allNodes[selectedNode].label = allNodes[selectedNode].hiddenLabel;
      allNodes[selectedNode].hiddenLabel = undefined;
    }
  } else if (highlightActive === true) {
    // console.log("highlightActive was true");
    // reset all nodes
    for (let nodeId in allNodes) {
      // allNodes[nodeId].color = "purple";
      allNodes[nodeId].color = nodeColors[nodeId];
      // delete allNodes[nodeId].color;
      if (allNodes[nodeId].hiddenLabel !== undefined) {
        allNodes[nodeId].label = allNodes[nodeId].hiddenLabel;
        allNodes[nodeId].hiddenLabel = undefined;
      }
    }
    highlightActive = false;
  }

  // transform the object into an array
  var updateArray = [];
  if (params.nodes.length > 0) {
    for (let nodeId in allNodes) {
      if (allNodes.hasOwnProperty(nodeId)) {
        // console.log(allNodes[nodeId]);
        updateArray.push(allNodes[nodeId]);
      }
    }
    nodes.update(updateArray);
  } else {
    // console.log("Nothing was selected");
    for (let nodeId in allNodes) {
      if (allNodes.hasOwnProperty(nodeId)) {
        // console.log(allNodes[nodeId]);
        // allNodes[nodeId].color = {};
        updateArray.push(allNodes[nodeId]);
      }
    }
    nodes.update(updateArray);
  }
}

function filterHighlight(params) {
  allNodes = nodes.get({ returnType: "Object" });
  // if something is selected:
  if (params.nodes.length > 0) {
    filterActive = true;
    let selectedNodes = params.nodes;

    // hiding all nodes and saving the label
    for (let nodeId in allNodes) {
      allNodes[nodeId].hidden = true;
      if (allNodes[nodeId].savedLabel === undefined) {
        allNodes[nodeId].savedLabel = allNodes[nodeId].label;
        allNodes[nodeId].label = undefined;
      }
    }

    for (let i=0; i < selectedNodes.length; i++) {
      allNodes[selectedNodes[i]].hidden = false;
      if (allNodes[selectedNodes[i]].savedLabel !== undefined) {
        allNodes[selectedNodes[i]].label = allNodes[selectedNodes[i]].savedLabel;
        allNodes[selectedNodes[i]].savedLabel = undefined;
      }
    }

  } else if (filterActive === true) {
    // reset all nodes
    for (let nodeId in allNodes) {
      allNodes[nodeId].hidden = false;
      if (allNodes[nodeId].savedLabel !== undefined) {
        allNodes[nodeId].label = allNodes[nodeId].savedLabel;
        allNodes[nodeId].savedLabel = undefined;
      }
    }
    filterActive = false;
  }

  // transform the object into an array
  var updateArray = [];
  if (params.nodes.length > 0) {
    for (let nodeId in allNodes) {
      if (allNodes.hasOwnProperty(nodeId)) {
        updateArray.push(allNodes[nodeId]);
      }
    }
    nodes.update(updateArray);
  } else {
    for (let nodeId in allNodes) {
      if (allNodes.hasOwnProperty(nodeId)) {
        updateArray.push(allNodes[nodeId]);
      }
    }
    nodes.update(updateArray);
  }
}

function selectNode(nodes) {
  network.selectNodes(nodes);
  neighbourhoodHighlight({ nodes: nodes });
  return nodes;
}

function selectNodes(nodes) {
  network.selectNodes(nodes);
  filterHighlight({nodes: nodes});
  return nodes;
}

function highlightFilter(filter) {
  let selectedNodes = []
  let selectedProp = filter['property']
  if (filter['item'] === 'node') {
    let allNodes = nodes.get({ returnType: "Object" });
    for (let nodeId in allNodes) {
      if (allNodes[nodeId][selectedProp] && filter['value'].includes((allNodes[nodeId][selectedProp]).toString())) {
        selectedNodes.push(nodeId)
      }
    }
  }
  else if (filter['item'] === 'edge'){
    let allEdges = edges.get({returnType: 'object'});
    // check if the selected property exists for selected edge and select the nodes connected to the edge
    for (let edge in allEdges) {
      if (allEdges[edge][selectedProp] && filter['value'].includes((allEdges[edge][selectedProp]).toString())) {
        selectedNodes.push(allEdges[edge]['from'])
        selectedNodes.push(allEdges[edge]['to'])
      }
    }
  }
  selectNodes(selectedNodes)
}</script>
            <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/vis-network/9.1.2/dist/dist/vis-network.min.css" integrity="sha512-WgxfT5LWjfszlPHXRmBWHkV2eceiWTOBvrKCNbdgDYTHrT2AeLCGbF4sZlZw3UMN3WtL0tGUoIAKsu8mllg/XA==" crossorigin="anonymous" referrerpolicy="no-referrer" />
            <script src="https://cdnjs.cloudflare.com/ajax/libs/vis-network/9.1.2/dist/vis-network.min.js" integrity="sha512-LnvoEWDFrqGHlHmDD2101OrLcbsfkrzoSpvtSQtxK3RMnRV0eOkhhBN2dXHKRrUU8p2DGRTk35n4O8nWSVe1mQ==" crossorigin="anonymous" referrerpolicy="no-referrer"></script>
            
            
            
            
            
            

        
<center>
<h1></h1>
</center>

<!-- <link rel="stylesheet" href="../node_modules/vis/dist/vis.min.css" type="text/css" />
<script type="text/javascript" src="../node_modules/vis/dist/vis.js"> </script>-->
        <link
          href="https://cdn.jsdelivr.net/npm/bootstrap@5.0.0-beta3/dist/css/bootstrap.min.css"
          rel="stylesheet"
          integrity="sha384-eOJMYsd53ii+scO/bJGFsiCZc+5NDVN2yr8+0RDqr0Ql0h+rP48ckxlpbzKgwra6"
          crossorigin="anonymous"
        />
        <script
          src="https://cdn.jsdelivr.net/npm/bootstrap@5.0.0-beta3/dist/js/bootstrap.bundle.min.js"
          integrity="sha384-JEW9xMcG8R+pH31jmWH6WWP0WintQrMb4s7ZOdauHnUtxwoG2vI5DkLtS3qm9Ekf"
          crossorigin="anonymous"
        ></script>


        <center>
          <h1></h1>
        </center>
        <style type="text/css">

             #mynetwork {
                 width: 100%;
                 height: 600px;
                 background-color: #ffffff;
                 border: 1px solid lightgray;
                 position: relative;
                 float: left;
             }

             
             #loadingBar {
                 position:absolute;
                 top:0px;
                 left:0px;
                 width: 100%;
                 height: 600px;
                 background-color:rgba(200,200,200,0.8);
                 -webkit-transition: all 0.5s ease;
                 -moz-transition: all 0.5s ease;
                 -ms-transition: all 0.5s ease;
                 -o-transition: all 0.5s ease;
                 transition: all 0.5s ease;
                 opacity:1;
             }

             #bar {
                 position:absolute;
                 top:0px;
                 left:0px;
                 width:20px;
                 height:20px;
                 margin:auto auto auto auto;
                 border-radius:11px;
                 border:2px solid rgba(30,30,30,0.05);
                 background: rgb(0, 173, 246); /* Old browsers */
                 box-shadow: 2px 0px 4px rgba(0,0,0,0.4);
             }

             #border {
                 position:absolute;
                 top:10px;
                 left:10px;
                 width:500px;
                 height:23px;
                 margin:auto auto auto auto;
                 box-shadow: 0px 0px 4px rgba(0,0,0,0.2);
                 border-radius:10px;
             }

             #text {
                 position:absolute;
                 top:8px;
                 left:530px;
                 width:30px;
                 height:50px;
                 margin:auto auto auto auto;
                 font-size:22px;
                 color: #000000;
             }

             div.outerBorder {
                 position:relative;
                 top:400px;
                 width:600px;
                 height:44px;
                 margin:auto auto auto auto;
                 border:8px solid rgba(0,0,0,0.1);
                 background: rgb(252,252,252); /* Old browsers */
                 background: -moz-linear-gradient(top,  rgba(252,252,252,1) 0%, rgba(237,237,237,1) 100%); /* FF3.6+ */
                 background: -webkit-gradient(linear, left top, left bottom, color-stop(0%,rgba(252,252,252,1)), color-stop(100%,rgba(237,237,237,1))); /* Chrome,Safari4+ */
                 background: -webkit-linear-gradient(top,  rgba(252,252,252,1) 0%,rgba(237,237,237,1) 100%); /* Chrome10+,Safari5.1+ */
                 background: -o-linear-gradient(top,  rgba(252,252,252,1) 0%,rgba(237,237,237,1) 100%); /* Opera 11.10+ */
                 background: -ms-linear-gradient(top,  rgba(252,252,252,1) 0%,rgba(237,237,237,1) 100%); /* IE10+ */
                 background: linear-gradient(to bottom,  rgba(252,252,252,1) 0%,rgba(237,237,237,1) 100%); /* W3C */
                 filter: progid:DXImageTransform.Microsoft.gradient( startColorstr='#fcfcfc', endColorstr='#ededed',GradientType=0 ); /* IE6-9 */
                 border-radius:72px;
                 box-shadow: 0px 0px 10px rgba(0,0,0,0.2);
             }
             

             

             
        </style>
    </head>


    <body>
        <div class="card" style="width: 100%">
            
            
            <div id="mynetwork" class="card-body"></div>
        </div>

        
            <div id="loadingBar">
              <div class="outerBorder">
                <div id="text">0%</div>
                <div id="border">
                  <div id="bar"></div>
                </div>
              </div>
            </div>
        
        

        <script type="text/javascript">

              // initialize global variables.
              var edges;
              var nodes;
              var allNodes;
              var allEdges;
              var nodeColors;
              var originalNodes;
              var network;
              var container;
              var options, data;
              var filter = {
                  item : '',
                  property : '',
                  value : []
              };

              

              

              // This method is responsible for drawing the graph, returns the drawn network
              function drawGraph() {
                  var container = document.getElementById('mynetwork');

                  

                  // parsing and collecting nodes and edges from the python
                  nodes = new vis.DataSet([{"color": "red", "id": "molecular modelling", "label": "molecular modelling", "shape": "dot"}, {"color": "#97c2fc", "id": "protein", "label": "protein", "shape": "dot"}, {"color": "#97c2fc", "id": "ligand", "label": "ligand", "shape": "dot"}, {"color": "#97c2fc", "id": "docking", "label": "docking", "shape": "dot"}, {"color": "#97c2fc", "id": "molecular dynamics", "label": "molecular dynamics", "shape": "dot"}, {"color": "#97c2fc", "id": "force field", "label": "force field", "shape": "dot"}, {"color": "#97c2fc", "id": "simulation", "label": "simulation", "shape": "dot"}, {"color": "#97c2fc", "id": "energy minimization", "label": "energy minimization", "shape": "dot"}, {"color": "#97c2fc", "id": "conformation", "label": "conformation", "shape": "dot"}, {"color": "#97c2fc", "id": "binding affinity", "label": "binding affinity", "shape": "dot"}, {"color": "#97c2fc", "id": "structure prediction", "label": "structure prediction", "shape": "dot"}, {"color": "#97c2fc", "id": "Receptor", "label": "Receptor", "shape": "dot"}, {"color": "#97c2fc", "id": "Docking", "label": "Docking", "shape": "dot"}, {"color": "#97c2fc", "id": "Binding Site", "label": "Binding Site", "shape": "dot"}, {"color": "#97c2fc", "id": "Pharmacophore", "label": "Pharmacophore", "shape": "dot"}, {"color": "#97c2fc", "id": "Dissociation Constant", "label": "Dissociation Constant", "shape": "dot"}, {"color": "#97c2fc", "id": "protein docking", "label": "protein docking", "shape": "dot"}, {"color": "#97c2fc", "id": "ligand docking", "label": "ligand docking", "shape": "dot"}, {"color": "#97c2fc", "id": "molecular docking", "label": "molecular docking", "shape": "dot"}, {"color": "#97c2fc", "id": "docking algorithm", "label": "docking algorithm", "shape": "dot"}, {"color": "#97c2fc", "id": "docking simulation", "label": "docking simulation", "shape": "dot"}, {"color": "#97c2fc", "id": "force fields", "label": "force fields", "shape": "dot"}, {"color": "#97c2fc", "id": "simulation software", "label": "simulation software", "shape": "dot"}, {"color": "#97c2fc", "id": "equations of motion", "label": "equations of motion", "shape": "dot"}, {"color": "#97c2fc", "id": "integrator algorithms", "label": "integrator algorithms", "shape": "dot"}, {"color": "#97c2fc", "id": "potential energy surface", "label": "potential energy surface", "shape": "dot"}, {"color": "#97c2fc", "id": "Lennard-Jones", "label": "Lennard-Jones", "shape": "dot"}, {"color": "#97c2fc", "id": "Electrostatics", "label": "Electrostatics", "shape": "dot"}, {"color": "#97c2fc", "id": "Van der Waals", "label": "Van der Waals", "shape": "dot"}, {"color": "#97c2fc", "id": "Nonbonded interactions", "label": "Nonbonded interactions", "shape": "dot"}, {"color": "#97c2fc", "id": "Atomistic potential", "label": "Atomistic potential", "shape": "dot"}, {"color": "#97c2fc", "id": "Monte Carlo simulation", "label": "Monte Carlo simulation", "shape": "dot"}, {"color": "#97c2fc", "id": "Brownian dynamics", "label": "Brownian dynamics", "shape": "dot"}, {"color": "#97c2fc", "id": "conformer", "label": "conformer", "shape": "dot"}, {"color": "#97c2fc", "id": "geometry optimization", "label": "geometry optimization", "shape": "dot"}, {"color": "#97c2fc", "id": "energy landscape", "label": "energy landscape", "shape": "dot"}, {"color": "#97c2fc", "id": "atomistic model", "label": "atomistic model", "shape": "dot"}, {"color": "#97c2fc", "id": "torsion angle", "label": "torsion angle", "shape": "dot"}, {"color": "#97c2fc", "id": "steric clashes", "label": "steric clashes", "shape": "dot"}, {"color": "#97c2fc", "id": "hydrogen bonding", "label": "hydrogen bonding", "shape": "dot"}, {"color": "#97c2fc", "id": "minimization", "label": "minimization", "shape": "dot"}, {"color": "#97c2fc", "id": "Ramachandran plot", "label": "Ramachandran plot", "shape": "dot"}, {"color": "#97c2fc", "id": "protein-ligand binding", "label": "protein-ligand binding", "shape": "dot"}, {"color": "#97c2fc", "id": "Kd value", "label": "Kd value", "shape": "dot"}, {"color": "#97c2fc", "id": "docking score", "label": "docking score", "shape": "dot"}, {"color": "#97c2fc", "id": "binding free energy", "label": "binding free energy", "shape": "dot"}, {"color": "#97c2fc", "id": "binding site interactions", "label": "binding site interactions", "shape": "dot"}, {"color": "#97c2fc", "id": "protein structure", "label": "protein structure", "shape": "dot"}, {"color": "#97c2fc", "id": "ligand binding", "label": "ligand binding", "shape": "dot"}, {"color": "#97c2fc", "id": "homology modeling", "label": "homology modeling", "shape": "dot"}, {"color": "#97c2fc", "id": "amino acid sequence", "label": "amino acid sequence", "shape": "dot"}, {"color": "#97c2fc", "id": "secondary structure", "label": "secondary structure", "shape": "dot"}, {"color": "#97c2fc", "id": "tertiary structure", "label": "tertiary structure", "shape": "dot"}, {"color": "#97c2fc", "id": "quaternary structure", "label": "quaternary structure", "shape": "dot"}, {"color": "#97c2fc", "id": "protein folding", "label": "protein folding", "shape": "dot"}, {"color": "#97c2fc", "id": "molecular mechanics", "label": "molecular mechanics", "shape": "dot"}, {"color": "#97c2fc", "id": "conformational search", "label": "conformational search", "shape": "dot"}, {"color": "#97c2fc", "id": "optimization algorithms", "label": "optimization algorithms", "shape": "dot"}, {"color": "#97c2fc", "id": "Protein-Ligand Docking", "label": "Protein-Ligand Docking", "shape": "dot"}, {"color": "#97c2fc", "id": "Molecular Docking", "label": "Molecular Docking", "shape": "dot"}, {"color": "#97c2fc", "id": "Virtual Screening", "label": "Virtual Screening", "shape": "dot"}, {"color": "#97c2fc", "id": "Ligand Binding", "label": "Ligand Binding", "shape": "dot"}, {"color": "#97c2fc", "id": "Active Site Prediction", "label": "Active Site Prediction", "shape": "dot"}, {"color": "#97c2fc", "id": "protein structure prediction", "label": "protein structure prediction", "shape": "dot"}, {"color": "#97c2fc", "id": "3D model building", "label": "3D model building", "shape": "dot"}, {"color": "#97c2fc", "id": "alignment of sequences", "label": "alignment of sequences", "shape": "dot"}, {"color": "#97c2fc", "id": "template-based modeling", "label": "template-based modeling", "shape": "dot"}, {"color": "#97c2fc", "id": "accuracy assessment", "label": "accuracy assessment", "shape": "dot"}, {"color": "#97c2fc", "id": "secondary structure prediction", "label": "secondary structure prediction", "shape": "dot"}, {"color": "#97c2fc", "id": "tertiary structure prediction", "label": "tertiary structure prediction", "shape": "dot"}, {"color": "#97c2fc", "id": "ab initio modeling", "label": "ab initio modeling", "shape": "dot"}, {"color": "#97c2fc", "id": "de novo modeling", "label": "de novo modeling", "shape": "dot"}, {"color": "#97c2fc", "id": "Protein structure", "label": "Protein structure", "shape": "dot"}, {"color": "#97c2fc", "id": "Ligand design", "label": "Ligand design", "shape": "dot"}, {"color": "#97c2fc", "id": "Macromolecular assembly", "label": "Macromolecular assembly", "shape": "dot"}, {"color": "#97c2fc", "id": "Molecular docking", "label": "Molecular docking", "shape": "dot"}, {"color": "#97c2fc", "id": "Visualization tools", "label": "Visualization tools", "shape": "dot"}, {"color": "#97c2fc", "id": "Sequence alignment", "label": "Sequence alignment", "shape": "dot"}, {"color": "#97c2fc", "id": "Homology modeling", "label": "Homology modeling", "shape": "dot"}, {"color": "#97c2fc", "id": "Multiple sequence alignment", "label": "Multiple sequence alignment", "shape": "dot"}, {"color": "#97c2fc", "id": "Structural superposition", "label": "Structural superposition", "shape": "dot"}, {"color": "#97c2fc", "id": "Comparative modeling", "label": "Comparative modeling", "shape": "dot"}, {"color": "#97c2fc", "id": "Threading", "label": "Threading", "shape": "dot"}, {"color": "#97c2fc", "id": "Protein structure prediction", "label": "Protein structure prediction", "shape": "dot"}, {"color": "#97c2fc", "id": "Model refinement", "label": "Model refinement", "shape": "dot"}, {"color": "#97c2fc", "id": "Structure alignment", "label": "Structure alignment", "shape": "dot"}, {"color": "#97c2fc", "id": "RMSD calculation", "label": "RMSD calculation", "shape": "dot"}, {"color": "#97c2fc", "id": "ligand binding affinity", "label": "ligand binding affinity", "shape": "dot"}, {"color": "#97c2fc", "id": "force field optimization", "label": "force field optimization", "shape": "dot"}, {"color": "#97c2fc", "id": "protein docking analysis", "label": "protein docking analysis", "shape": "dot"}, {"color": "#97c2fc", "id": "Root Mean Square Deviation", "label": "Root Mean Square Deviation", "shape": "dot"}, {"color": "#97c2fc", "id": "Structural Alignment", "label": "Structural Alignment", "shape": "dot"}, {"color": "#97c2fc", "id": "Protein Conformation", "label": "Protein Conformation", "shape": "dot"}, {"color": "#97c2fc", "id": "Structural Changes", "label": "Structural Changes", "shape": "dot"}, {"color": "#97c2fc", "id": "Ab initio prediction", "label": "Ab initio prediction", "shape": "dot"}, {"color": "#97c2fc", "id": "Fold recognition", "label": "Fold recognition", "shape": "dot"}, {"color": "#97c2fc", "id": "Protein structure refinement", "label": "Protein structure refinement", "shape": "dot"}, {"color": "#97c2fc", "id": "protein-ligand interactions", "label": "protein-ligand interactions", "shape": "dot"}, {"color": "#97c2fc", "id": "docking energy", "label": "docking energy", "shape": "dot"}, {"color": "#97c2fc", "id": "binding site", "label": "binding site", "shape": "dot"}, {"color": "#97c2fc", "id": "binding affinity prediction", "label": "binding affinity prediction", "shape": "dot"}, {"color": "#97c2fc", "id": "scoring function", "label": "scoring function", "shape": "dot"}, {"color": "#97c2fc", "id": "parameter optimization", "label": "parameter optimization", "shape": "dot"}, {"color": "#97c2fc", "id": "molecular dynamics simulations", "label": "molecular dynamics simulations", "shape": "dot"}, {"color": "#97c2fc", "id": "steepest descent", "label": "steepest descent", "shape": "dot"}, {"color": "#97c2fc", "id": "genetic algorithm", "label": "genetic algorithm", "shape": "dot"}, {"color": "#97c2fc", "id": "protein binding", "label": "protein binding", "shape": "dot"}, {"color": "#97c2fc", "id": "molecular interaction", "label": "molecular interaction", "shape": "dot"}, {"color": "#97c2fc", "id": "protein-protein interactions", "label": "protein-protein interactions", "shape": "dot"}, {"color": "#97c2fc", "id": "Protein-protein interactions", "label": "Protein-protein interactions", "shape": "dot"}, {"color": "#97c2fc", "id": "Binding affinities", "label": "Binding affinities", "shape": "dot"}, {"color": "#97c2fc", "id": "Binding pockets", "label": "Binding pockets", "shape": "dot"}, {"color": "#97c2fc", "id": "enzyme-substrate interaction", "label": "enzyme-substrate interaction", "shape": "dot"}, {"color": "#97c2fc", "id": "protein-protein interaction", "label": "protein-protein interaction", "shape": "dot"}, {"color": "#97c2fc", "id": "ligand-receptor binding", "label": "ligand-receptor binding", "shape": "dot"}, {"color": "#97c2fc", "id": "Protein-ligand interactions", "label": "Protein-ligand interactions", "shape": "dot"}, {"color": "#97c2fc", "id": "Molecular docking software", "label": "Molecular docking software", "shape": "dot"}, {"color": "#97c2fc", "id": "Virtual screening", "label": "Virtual screening", "shape": "dot"}, {"color": "#97c2fc", "id": "Binding affinity", "label": "Binding affinity", "shape": "dot"}, {"color": "#97c2fc", "id": "Scoring functions", "label": "Scoring functions", "shape": "dot"}, {"color": "#97c2fc", "id": "binding energy", "label": "binding energy", "shape": "dot"}, {"color": "#97c2fc", "id": "ligand interactions", "label": "ligand interactions", "shape": "dot"}, {"color": "#97c2fc", "id": "receptor conformation", "label": "receptor conformation", "shape": "dot"}, {"color": "#97c2fc", "id": "hydrophobic interactions", "label": "hydrophobic interactions", "shape": "dot"}, {"color": "#97c2fc", "id": "polar interactions", "label": "polar interactions", "shape": "dot"}, {"color": "#97c2fc", "id": "protein-protein docking", "label": "protein-protein docking", "shape": "dot"}, {"color": "#97c2fc", "id": "protein-protein binding", "label": "protein-protein binding", "shape": "dot"}, {"color": "#97c2fc", "id": "protein-protein interface", "label": "protein-protein interface", "shape": "dot"}, {"color": "#97c2fc", "id": "protein-protein complexes", "label": "protein-protein complexes", "shape": "dot"}, {"color": "#97c2fc", "id": "protein-protein analysis", "label": "protein-protein analysis", "shape": "dot"}]);
                  edges = new vis.DataSet([{"from": "molecular modelling", "title": "Similarity: 0.79", "to": "protein"}, {"from": "molecular modelling", "title": "Similarity: 0.82", "to": "ligand"}, {"from": "molecular modelling", "title": "Similarity: 0.83", "to": "docking"}, {"from": "molecular modelling", "title": "Similarity: 0.9", "to": "molecular dynamics"}, {"from": "molecular modelling", "title": "Similarity: 0.78", "to": "force field"}, {"from": "molecular modelling", "title": "Similarity: 0.83", "to": "simulation"}, {"from": "molecular modelling", "title": "Similarity: 0.79", "to": "energy minimization"}, {"from": "molecular modelling", "title": "Similarity: 0.83", "to": "conformation"}, {"from": "molecular modelling", "title": "Similarity: 0.79", "to": "binding affinity"}, {"from": "molecular modelling", "title": "Similarity: 0.81", "to": "structure prediction"}, {"from": "ligand", "title": "Similarity: 0.86", "to": "Receptor"}, {"from": "ligand", "title": "Similarity: 0.81", "to": "Docking"}, {"from": "ligand", "title": "Similarity: 0.84", "to": "Binding Site"}, {"from": "ligand", "title": "Similarity: 0.83", "to": "Pharmacophore"}, {"from": "ligand", "title": "Similarity: 0.78", "to": "Dissociation Constant"}, {"from": "docking", "title": "Similarity: 0.93", "to": "protein docking"}, {"from": "docking", "title": "Similarity: 0.91", "to": "ligand docking"}, {"from": "docking", "title": "Similarity: 0.92", "to": "molecular docking"}, {"from": "docking", "title": "Similarity: 0.93", "to": "docking algorithm"}, {"from": "docking", "title": "Similarity: 0.94", "to": "docking simulation"}, {"from": "molecular dynamics", "title": "Similarity: 0.8", "to": "force fields"}, {"from": "molecular dynamics", "title": "Similarity: 0.81", "to": "simulation software"}, {"from": "molecular dynamics", "title": "Similarity: 0.82", "to": "equations of motion"}, {"from": "molecular dynamics", "title": "Similarity: 0.76", "to": "integrator algorithms"}, {"from": "molecular dynamics", "title": "Similarity: 0.79", "to": "potential energy surface"}, {"from": "force field", "title": "Similarity: 0.79", "to": "Lennard-Jones"}, {"from": "force field", "title": "Similarity: 0.79", "to": "Electrostatics"}, {"from": "force field", "title": "Similarity: 0.78", "to": "Van der Waals"}, {"from": "force field", "title": "Similarity: 0.78", "to": "Nonbonded interactions"}, {"from": "force field", "title": "Similarity: 0.79", "to": "Atomistic potential"}, {"from": "simulation", "title": "Similarity: 0.81", "to": "molecular dynamics"}, {"from": "simulation", "title": "Similarity: 0.85", "to": "Monte Carlo simulation"}, {"from": "simulation", "title": "Similarity: 0.8", "to": "force fields"}, {"from": "simulation", "title": "Similarity: 0.79", "to": "energy minimization"}, {"from": "simulation", "title": "Similarity: 0.8", "to": "Brownian dynamics"}, {"from": "energy minimization", "title": "Similarity: 0.78", "to": "force field"}, {"from": "energy minimization", "title": "Similarity: 0.78", "to": "conformer"}, {"from": "energy minimization", "title": "Similarity: 0.83", "to": "geometry optimization"}, {"from": "energy minimization", "title": "Similarity: 0.86", "to": "energy landscape"}, {"from": "energy minimization", "title": "Similarity: 0.77", "to": "atomistic model"}, {"from": "energy minimization", "title": "Similarity: 0.79", "to": "molecular mechanics"}, {"from": "energy minimization", "title": "Similarity: 0.79", "to": "conformational search"}, {"from": "energy minimization", "title": "Similarity: 0.83", "to": "optimization algorithms"}, {"from": "energy minimization", "title": "Similarity: 0.84", "to": "potential energy surface"}, {"from": "conformation", "title": "Similarity: 0.8", "to": "torsion angle"}, {"from": "conformation", "title": "Similarity: 0.81", "to": "steric clashes"}, {"from": "conformation", "title": "Similarity: 0.81", "to": "hydrogen bonding"}, {"from": "conformation", "title": "Similarity: 0.8", "to": "minimization"}, {"from": "conformation", "title": "Similarity: 0.77", "to": "Ramachandran plot"}, {"from": "binding affinity", "title": "Similarity: 0.86", "to": "protein-ligand binding"}, {"from": "binding affinity", "title": "Similarity: 0.76", "to": "Kd value"}, {"from": "binding affinity", "title": "Similarity: 0.78", "to": "docking score"}, {"from": "binding affinity", "title": "Similarity: 0.87", "to": "binding free energy"}, {"from": "binding affinity", "title": "Similarity: 0.88", "to": "binding site interactions"}, {"from": "structure prediction", "title": "Similarity: 0.82", "to": "protein structure"}, {"from": "structure prediction", "title": "Similarity: 0.76", "to": "ligand binding"}, {"from": "structure prediction", "title": "Similarity: 0.75", "to": "energy minimization"}, {"from": "structure prediction", "title": "Similarity: 0.78", "to": "docking simulation"}, {"from": "structure prediction", "title": "Similarity: 0.81", "to": "homology modeling"}, {"from": "ligand docking", "title": "Similarity: 0.89", "to": "Protein-ligand interactions"}, {"from": "ligand docking", "title": "Similarity: 0.91", "to": "Molecular docking software"}, {"from": "ligand docking", "title": "Similarity: 0.8", "to": "Virtual screening"}, {"from": "ligand docking", "title": "Similarity: 0.83", "to": "Binding affinity"}, {"from": "ligand docking", "title": "Similarity: 0.75", "to": "Scoring functions"}, {"from": "docking simulation", "title": "Similarity: 0.87", "to": "Protein-Ligand Docking"}, {"from": "docking simulation", "title": "Similarity: 0.9", "to": "Molecular Docking"}, {"from": "docking simulation", "title": "Similarity: 0.81", "to": "Virtual Screening"}, {"from": "docking simulation", "title": "Similarity: 0.81", "to": "Ligand Binding"}, {"from": "docking simulation", "title": "Similarity: 0.8", "to": "Active Site Prediction"}, {"from": "protein structure", "title": "Similarity: 0.84", "to": "amino acid sequence"}, {"from": "protein structure", "title": "Similarity: 0.87", "to": "secondary structure"}, {"from": "protein structure", "title": "Similarity: 0.86", "to": "tertiary structure"}, {"from": "protein structure", "title": "Similarity: 0.85", "to": "quaternary structure"}, {"from": "protein structure", "title": "Similarity: 0.89", "to": "protein folding"}, {"from": "homology modeling", "title": "Similarity: 0.83", "to": "protein structure prediction"}, {"from": "homology modeling", "title": "Similarity: 0.83", "to": "3D model building"}, {"from": "homology modeling", "title": "Similarity: 0.8", "to": "alignment of sequences"}, {"from": "homology modeling", "title": "Similarity: 0.87", "to": "template-based modeling"}, {"from": "homology modeling", "title": "Similarity: 0.74", "to": "accuracy assessment"}, {"from": "protein structure prediction", "title": "Similarity: 0.92", "to": "secondary structure prediction"}, {"from": "protein structure prediction", "title": "Similarity: 0.91", "to": "tertiary structure prediction"}, {"from": "protein structure prediction", "title": "Similarity: 0.82", "to": "ab initio modeling"}, {"from": "protein structure prediction", "title": "Similarity: 0.79", "to": "de novo modeling"}, {"from": "protein structure prediction", "title": "Similarity: 0.84", "to": "Homology modeling"}, {"from": "protein structure prediction", "title": "Similarity: 0.86", "to": "Ab initio prediction"}, {"from": "protein structure prediction", "title": "Similarity: 0.78", "to": "Fold recognition"}, {"from": "protein structure prediction", "title": "Similarity: 0.72", "to": "Threading"}, {"from": "protein structure prediction", "title": "Similarity: 0.9", "to": "Protein structure refinement"}, {"from": "3D model building", "title": "Similarity: 0.79", "to": "Protein structure"}, {"from": "3D model building", "title": "Similarity: 0.79", "to": "Ligand design"}, {"from": "3D model building", "title": "Similarity: 0.79", "to": "Macromolecular assembly"}, {"from": "3D model building", "title": "Similarity: 0.8", "to": "Molecular docking"}, {"from": "3D model building", "title": "Similarity: 0.82", "to": "Visualization tools"}, {"from": "alignment of sequences", "title": "Similarity: 0.93", "to": "Sequence alignment"}, {"from": "alignment of sequences", "title": "Similarity: 0.8", "to": "Homology modeling"}, {"from": "alignment of sequences", "title": "Similarity: 0.93", "to": "Multiple sequence alignment"}, {"from": "alignment of sequences", "title": "Similarity: 0.78", "to": "Structural superposition"}, {"from": "alignment of sequences", "title": "Similarity: 0.76", "to": "Comparative modeling"}, {"from": "template-based modeling", "title": "Similarity: 0.87", "to": "Homology modeling"}, {"from": "template-based modeling", "title": "Similarity: 0.8", "to": "Threading"}, {"from": "template-based modeling", "title": "Similarity: 0.78", "to": "Protein structure prediction"}, {"from": "template-based modeling", "title": "Similarity: 0.85", "to": "Model refinement"}, {"from": "template-based modeling", "title": "Similarity: 0.79", "to": "Structure alignment"}, {"from": "accuracy assessment", "title": "Similarity: 0.77", "to": "RMSD calculation"}, {"from": "accuracy assessment", "title": "Similarity: 0.74", "to": "protein structure prediction"}, {"from": "accuracy assessment", "title": "Similarity: 0.74", "to": "ligand binding affinity"}, {"from": "accuracy assessment", "title": "Similarity: 0.75", "to": "force field optimization"}, {"from": "accuracy assessment", "title": "Similarity: 0.75", "to": "protein docking analysis"}, {"from": "RMSD calculation", "title": "Similarity: 0.83", "to": "Root Mean Square Deviation"}, {"from": "RMSD calculation", "title": "Similarity: 0.75", "to": "Structural Alignment"}, {"from": "RMSD calculation", "title": "Similarity: 0.8", "to": "Protein Conformation"}, {"from": "RMSD calculation", "title": "Similarity: 0.73", "to": "Structural Changes"}, {"from": "RMSD calculation", "title": "Similarity: 0.84", "to": "Molecular Docking"}, {"from": "ligand binding affinity", "title": "Similarity: 0.87", "to": "protein-ligand interactions"}, {"from": "ligand binding affinity", "title": "Similarity: 0.82", "to": "docking energy"}, {"from": "ligand binding affinity", "title": "Similarity: 0.85", "to": "binding site"}, {"from": "ligand binding affinity", "title": "Similarity: 0.89", "to": "binding affinity prediction"}, {"from": "ligand binding affinity", "title": "Similarity: 0.74", "to": "scoring function"}, {"from": "force field optimization", "title": "Similarity: 0.82", "to": "energy minimization"}, {"from": "force field optimization", "title": "Similarity: 0.86", "to": "parameter optimization"}, {"from": "force field optimization", "title": "Similarity: 0.79", "to": "molecular dynamics simulations"}, {"from": "force field optimization", "title": "Similarity: 0.76", "to": "steepest descent"}, {"from": "force field optimization", "title": "Similarity: 0.79", "to": "genetic algorithm"}, {"from": "protein docking analysis", "title": "Similarity: 0.85", "to": "protein binding"}, {"from": "protein docking analysis", "title": "Similarity: 0.81", "to": "molecular interaction"}, {"from": "protein docking analysis", "title": "Similarity: 0.91", "to": "ligand docking"}, {"from": "protein docking analysis", "title": "Similarity: 0.88", "to": "docking energy"}, {"from": "protein docking analysis", "title": "Similarity: 0.85", "to": "protein-protein interactions"}, {"from": "docking energy", "title": "Similarity: 0.89", "to": "binding energy"}, {"from": "docking energy", "title": "Similarity: 0.83", "to": "ligand interactions"}, {"from": "docking energy", "title": "Similarity: 0.81", "to": "receptor conformation"}, {"from": "docking energy", "title": "Similarity: 0.8", "to": "hydrophobic interactions"}, {"from": "docking energy", "title": "Similarity: 0.81", "to": "polar interactions"}, {"from": "protein binding", "title": "Similarity: 0.87", "to": "Protein-protein interactions"}, {"from": "protein binding", "title": "Similarity: 0.87", "to": "Binding affinities"}, {"from": "protein binding", "title": "Similarity: 0.84", "to": "Molecular docking"}, {"from": "protein binding", "title": "Similarity: 0.79", "to": "Homology modeling"}, {"from": "protein binding", "title": "Similarity: 0.84", "to": "Binding pockets"}, {"from": "molecular interaction", "title": "Similarity: 0.85", "to": "protein-ligand binding"}, {"from": "molecular interaction", "title": "Similarity: 0.87", "to": "enzyme-substrate interaction"}, {"from": "molecular interaction", "title": "Similarity: 0.9", "to": "protein-protein interaction"}, {"from": "molecular interaction", "title": "Similarity: 0.85", "to": "ligand-receptor binding"}, {"from": "molecular interaction", "title": "Similarity: 0.9", "to": "molecular docking"}, {"from": "protein-protein interactions", "title": "Similarity: 0.92", "to": "protein-protein docking"}, {"from": "protein-protein interactions", "title": "Similarity: 0.94", "to": "protein-protein binding"}, {"from": "protein-protein interactions", "title": "Similarity: 0.96", "to": "protein-protein interface"}, {"from": "protein-protein interactions", "title": "Similarity: 0.95", "to": "protein-protein complexes"}, {"from": "protein-protein interactions", "title": "Similarity: 0.91", "to": "protein-protein analysis"}]);

                  nodeColors = {};
                  allNodes = nodes.get({ returnType: "Object" });
                  for (nodeId in allNodes) {
                    nodeColors[nodeId] = allNodes[nodeId].color;
                  }
                  allEdges = edges.get({ returnType: "Object" });
                  // adding nodes and edges to the graph
                  data = {nodes: nodes, edges: edges};

                  var options = {
    "configure": {
        "enabled": false
    },
    "edges": {
        "color": {
            "inherit": true
        },
        "smooth": {
            "enabled": true,
            "type": "dynamic"
        }
    },
    "interaction": {
        "dragNodes": true,
        "hideEdgesOnDrag": false,
        "hideNodesOnDrag": false
    },
    "physics": {
        "enabled": true,
        "stabilization": {
            "enabled": true,
            "fit": true,
            "iterations": 1000,
            "onlyDynamicEdges": false,
            "updateInterval": 50
        }
    }
};

                  


                  

                  network = new vis.Network(container, data, options);

                  

                  

                  


                  
                      network.on("stabilizationProgress", function(params) {
                          document.getElementById('loadingBar').removeAttribute("style");
                          var maxWidth = 496;
                          var minWidth = 20;
                          var widthFactor = params.iterations/params.total;
                          var width = Math.max(minWidth,maxWidth * widthFactor);
                          document.getElementById('bar').style.width = width + 'px';
                          document.getElementById('text').innerHTML = Math.round(widthFactor*100) + '%';
                      });
                      network.once("stabilizationIterationsDone", function() {
                          document.getElementById('text').innerHTML = '100%';
                          document.getElementById('bar').style.width = '496px';
                          document.getElementById('loadingBar').style.opacity = 0;
                          // really clean the dom element
                          setTimeout(function () {document.getElementById('loadingBar').style.display = 'none';}, 500);
                      });
                  

                  return network;

              }
              drawGraph();
        </script>
    </body>
</html>